internal/controller/genericprovider_controller.go

```diff
@@ -39,20 +42,28 @@ import (
 )
 
 type GenericProviderReconciler struct {
-	Provider     genericprovider.GenericProvider
-	ProviderList genericprovider.GenericProviderList
-	Client       client.Client
-	Config       *rest.Config
+	Provider                 genericprovider.GenericProvider
+	ProviderList             genericprovider.GenericProviderList
+	Client                   client.Client
+	Config                   *rest.Config
+	WatchConfigSecretChanges bool
 }
 
 const (
 	appliedSpecHashAnnotation = "operator.cluster.x-k8s.io/applied-spec-hash"
 )
 
 func (r *GenericProviderReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(r.Provider).
-		WithOptions(options).
+	builder := ctrl.NewControllerManagedBy(mgr).
+		For(r.Provider)
+	if r.WatchConfigSecretChanges {
+		builder = builder.Watches(&corev1.Secret{}, &providerSecretMapper{
```

Comment:
It may be better to use `handler.MapFunc`, similar to this, since using queue directly is “internal-ish” way of doing that.
```go
builder.Watch(
	source.Kind(mgr.GetCache(), &corev1.Secret{}),
	handler.EnqueueRequestsFromMapFunc(secretToProviders),
)
// where 
func secretToProviders(ctx context.Context, secret client.Object) []ctrl.Request
// loop over listConfigSecretUsers
```

Reply:
I addressed your comment.
Apparently, `builder` does not have a `Watch` function but a `Watches` one which accepts a plain object as first parameter.
I though used `handler.EnqueueRequestsFromMapFunc` as suggested and adapted the rest of the code